File: P1.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 27 13:02:02 2021

@author: jgdch
"""

#Time Compleixty:O(N**2), N is the number of ndoes in the tree
#Space Complexity:O(N)
#Passed all LC testcases succesfully
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    def helper(self, node, remainingSum, path, paths):
        
        if not node:
            return 
        
        path.append(node.val)
        
        if remainingSum == node.val and not node.left and not node.right:
            paths.append(list(path))
        else:    
            self.helper(node.left, remainingSum - node.val, path, paths)
            self.helper(node.right, remainingSum - node.val, path, paths)
            
        path.pop()    
    
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        paths = []
        self.helper(root, sum, [], paths)
        return paths        

File: p1.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //Declaring result in global scope
    List<List<Integer>> result;
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        //Initialize result
        result = new ArrayList<>();
        //Using arraylist to store the path and integer to store the currentsum
        List<Integer> list = new ArrayList<>();
        helper(root, 0, list, targetSum );
        return result;
    }

    public void helper(TreeNode root,int csum, List<Integer> list, int targetSum ){
        //base condition
        //Add the current value to sum and if that matches the target then create a new list and put elements of the list into it. Add new list to result
        if(root == null) return;
        csum = csum + root.val;
        list.add(root.val);
        if(root.left == null && root.right == null){
            if(csum == targetSum){
                List<Integer> res = new ArrayList<>(list);
                result.add(res);
                
            }
            
        }

        //left side recursion
        helper(root.left, csum, list, targetSum );

        //right side recursion
        helper(root.right, csum, list, targetSum );

        //backtracking the element as list is passed by reference
        list.remove(list.size() - 1);

    }
}